### PR TITLE
admin: add API for creating chunked handlers

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: The admin handler structure uses a single callback that is expected to fill in the entire response body, making it impossible to stream data out as it is generated. This PR refactors the internal admin structures to require a Handler object with a nextChunk method, to enable chunking. It defers to other PRs to
 * actually convert any handlers to use chunking (e.g. https://github.com/envoyproxy/envoy/pull/19693)
 * implement proper flow control in the AdminFilter to avoid generating chunks before clients are ready to receive them (https://github.com/envoyproxy/envoy/pull/19898)
Additional Description:
Risk Level: medium -- this is a pure refactor but not a trivial one. The same exact admin content should be generated before/after this change
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/server/admin/admin.cc [hunk 1]
```diff
@@ -161,63 +161,67 @@ AdminImpl::AdminImpl(const std::string& profile_path, Server::Instance& server,
...
-           MAKE_ADMIN_HANDLER(stats_handler_.handlerStatsRecentLookupsEnable), false, true},
-          {"/listeners", "print listener info",
-           MAKE_ADMIN_HANDLER(listeners_handler_.handlerListenerInfo), false, false},
-          {"/runtime", "print runtime values", MAKE_ADMIN_HANDLER(runtime_handler_.handlerRuntime),
-           false, false},
-          {"/runtime_modify", "modify runtime values",
-           MAKE_ADMIN_HANDLER(runtime_handler_.handlerRuntimeModify), false, true},
-          {"/reopen_logs", "reopen access logs",
-           MAKE_ADMIN_HANDLER(logs_handler_.handlerReopenLogs), false, true},
+          makeHandler("/", "Admin home page", MAKE_ADMIN_HANDLER(handlerAdminHome), false, false),
```
Comment (jmarantz) [inline source/server/admin/admin.cc:164]: note this block of code is mechanistically transformed from its previous state, as we needed to add Handler wrappers to all the legacy callback-based handlers.

In a future PR, `/stats` and likely also `/cluster_info` will be converted to Chunked handlers.

**Comment by jmarantz** [comment]
@krajshiva FYI

File: source/server/admin/admin.cc [hunk 2]
```diff
@@ -247,49 +251,108 @@ bool AdminImpl::createNetworkFilterChain(Network::Connection& connection,
...
+
+Admin::HandlerPtr AdminImpl::makeStaticTextHandler(absl::string_view response, Http::Code code) {
+  return std::make_unique<StaticTextHandler>(response, code);
 }
 
 Http::Code AdminImpl::runCallback(absl::string_view path_and_query,
                                   Http::ResponseHeaderMap& response_headers,
                                   Buffer::Instance& response, AdminStream& admin_stream) {
+  HandlerPtr handler = findHandler(path_and_query, admin_stream);
+  Http::Code code = handler->start(/*path_and_query, */ response_headers);
```
Comment (zuercher) [inline source/server/admin/admin.cc:318]: Remove the `path_and_query` comment?

File: source/server/admin/admin_filter.cc [hunk 3]
```diff
@@ -65,18 +65,25 @@ void AdminFilter::onComplete() {
   const absl::string_view path = request_headers_->getPathValue();
   ENVOY_STREAM_LOG(debug, "request complete: path: {}", *decoder_callbacks_, path);
 
-  Buffer::OwnedImpl response;
   auto header_map = Http::ResponseHeaderMapImpl::create();
   RELEASE_ASSERT(request_headers_, "");
-  Http::Code code = admin_server_callback_func_(path, *header_map, response, *this);
+  Admin::HandlerPtr handler = admin_handler_fn_(path, *this);
+  Http::Code code = handler->start(/*path, */ *header_map);
```
Comment (zuercher) [inline source/server/admin/admin_filter.cc:71]: Remove path comment?

File: envoy/server/admin.h [hunk 4]
```diff
@@ -76,6 +76,34 @@ class Admin {
...
+     *        those here.
+     * @return the HTTP status of the response.
+     */
+    virtual Http::Code start(Http::ResponseHeaderMap& response_headers) PURE;
+
+    /**
+     * Adds the next chunk of data to the response. Note that nextChunk can
+     * return 'true' but not add any data to the response, in which case a chunk
+     * is not sent, and a subsequent call to nextChunk can be made later,
+     * possibly after a post() or low-watermark callback on the http filter.
```
Comment (zuercher) [inline envoy/server/admin.h:98]: I gather this is only true once #19898 lands and until then we spin calling nextChunk. I think it's ok to leave as long as you expect that PR to land soon-ish.

**Review by zuercher — APPROVED** [review]
Looks good. Couple small comments, but they could be addressed in one of your follow-up PRs, if you'd like.

**Comment by jmarantz** [comment]
Thanks -- that's right on all counts and I'll clean up the comments in the next PR.

I think the next PR is working but I can't prove it yet; I think I need a test using HttpAsyncHandler or something.

**Pull request merged**